api/enki.py

```diff
@@ -0,0 +1,556 @@
+    #TODO
+    """Fill in bibliographic metadata for Enki records.
+
+    Currently this is only used by BibliographicRefreshScript. It's
+    not normally necessary because the Enki API combines
+    bibliographic and availability data.
+    """
+    def __init__(self, _db, metadata_replacement_policy=None, enki_api=None,
+                 input_identifier_types=None, input_identifiers=None, **kwargs):
+        """
+        :param input_identifier_types: Passed in by RunCoverageProviderScript, data sources to get coverage for.
+        :param input_identifiers: Passed in by RunCoverageProviderScript, specific identifiers to get coverage for.
+        """
+        self.parser = BibliographicParser()
+        super(EnkiBibliographicCoverageProvider, self).__init__(
+            _db, enki_api, DataSource.ENKI,
+            batch_size=25,
+            metadata_replacement_policy=metadata_replacement_policy,
+            **kwargs
+        )
+
+    def process_batch(self, identifiers):
+        identifier_strings = self.api.create_identifier_strings(identifiers)
+        response = self.api.availability(title_ids=identifier_strings)
+        seen_identifiers = set()
+        batch_results = []
+        for metadata, availability in self.parser.process_all(response.content):
+            identifier, is_new = metadata.primary_identifier.load(self._db)
+            if not identifier in identifiers:
+                # Enki told us about a book we didn't ask
+                # for. This shouldn't happen, but if it does we should
+                # do nothing further.
+                continue
+            seen_identifiers.add(identifier.identifier)
+            result = self.set_metadata(identifier, metadata)
+            if not isinstance(result, CoverageFailure):
+                result = self.handle_success(identifier)
+            batch_results.append(result)
+
+        # Create a CoverageFailure object for each original identifier
+        # not mentioned in the results.
+        for identifier_string in identifier_strings:
+            if identifier_string not in seen_identifiers:
+                identifier, ignore = Identifier.for_foreign_id(
+                    self._db, Identifier.ENKI_ID, identifier_string
+                )
+                result = CoverageFailure(
+                    identifier, "Book not found in Enki", data_source=self.output_source, transient=False
+                )
+                batch_results.append(result)
+        return batch_results
+
+    def handle_success(self, identifier):
+        return self.set_presentation_ready(identifier)
+
+    def process_item(self, identifier):
+        results = self.process_batch([identifier])
+        return results[0]
+
+class BibliographicParser(object):
+
+    """Helper function to parse JSON"""
+    def process_all(self, json_data):
+        data = json.loads(json_data)
+	returned_titles = data["result"]["titles"]
```

Comment:
I don't know how big a problem this will cause, but there's an indentation mismatch here.